src/FormElementManager.php

```diff
@@ -32,10 +33,8 @@
  *
  * Enforces that elements retrieved are instances of ElementInterface.
  */
-class FormElementManagerV3Polyfill extends AbstractPluginManager
```

Comment:
Isn't renaming a class a BC break?

Reply:
I do not see the name as a problem because an alias was created:

https://github.com/laminas/laminas-form/blob/a028e927ed1b07a98e43c22d5e7cac796f7a0e9d/autoload/formElementManagerPolyfill.php#L12-L18

But the polyfill classes and the related trait are not marked as internal and can be overwritten in userland code. For a smoothly transition we need a deprecation – in code and documentation.

@Slamdunk 

The following steps are needed:

* create a new class for `Laminas\Form\FormElementManager`
* add deprecation notices to the old trait und the old polyfill classes under the namespace `Laminas\Form\FormElementManager`
* add a migration guide to the documentation and add the deprecation infos (an example for deprecations can be found on [Mezzio](https://docs.mezzio.dev/mezzio/v2/reference/migration-to-v2-2/))

Reply:
I don't think we need a transition.

Yes: at raw level this class has never been marked as `@internal`, so it may appear as a BC Break.

But both `FormElementManagerV2Polyfill` and `FormElementManagerV3Polyfill` are never documented, type-hinted nor used as-is since they have always be used only once aliased:

https://github.com/laminas/laminas-form/blob/a028e927ed1b07a98e43c22d5e7cac796f7a0e9d/src/ConfigProvider.php#L42

https://github.com/laminas/laminas-form/blob/a028e927ed1b07a98e43c22d5e7cac796f7a0e9d/src/Factory.php#L92

To me it's just a bug related to the missing `@internal`, a best practice introduced recently; when those polyfills where created no one used that annotation.

Reply:
> Yes: at raw level this class has never been marked as `@internal`, so it may appear as a BC Break.

Correct, the marking of the class is missing therefore the problem is in this component and cannot be at the user's expense.

> But both `FormElementManagerV2Polyfill` and `FormElementManagerV3Polyfill` are never documented…

Unfortunately we cannot rely on this fact because the entire documentation (code and docs) are not good for laminas-form.

> …type-hinted nor used as-is since they have always be used only once aliased…

Because of the class alias you will find the following use in applications:

```php
/** @var \Laminas\Form\FormElementManager|\Laminas\Form\FormElementManager\FormElementManagerV3Polyfill $formElementManager */
$formElementManager = $container->get('FormElementManager');
```

This will not break an application if the polyfill is removed but an IDE can highlight a deprecation notice.

> To me it's just a bug related to the missing `@internal`…

Yes it is a bug but not the fault of the user.

---

Please take the necessary steps to ensure that the rules of semantic versioning are followed. Thanks in advance!

Reply:
@Slamdunk generally I agree with you. it _should_ have been marked as `@internal`, but it wasn't. so you can assume, that some users use the class directly, e.g. myself 🙂
we extend this V3 polyfill with a dummy class to be able to type hint to a class and allowing auto-wiring.
For my projects it's not a problem, I can resolve such a case pretty quickly because I'm quite familiar with laminas code. but probably lots of consumers aren't and will be annoyed for having to fix a bc break in a minor version (even though if they extend such a weird polyfill class they have it coming to them 😋).
on the other side it doesn't hurt to keep the code and mark it as deprecated.
I was having the same problem in laminas-session, you can [look it up there](https://github.com/laminas/laminas-session/pull/35). TL;DR: I copied the polyfill code to the new class so that they aren't used anymore and marked them as deprecated.
The V2 polyfill was deleted because it can't be used anymore. it's not a bc break in that case.

Reply:
> generally I agree with you.

The same here. It would be great if the obsolete classes could simply be removed, but now the component must deal with the problem.

Reply:
>  I was having the same problem in laminas-session, you can [look it up there](https://github.com/laminas/laminas-session/pull/35). TL;DR: I copied the polyfill code to the new class so that they aren't used anymore and marked them as deprecated.
> The V2 polyfill was deleted because it can't be used anymore. it's not a bc break in that case.

Correct me if I'm wrong: by the same logic, stated that in this PR I've removed support for SMv2, I can rearrange the steps asked by @froschdesign in:

1. Keep `2.16` version of `FormElementManagerTrait` and mark it as `@deprected` and unmaintained
1. Delete `FormElementManagerV2Polyfill` as unusable anyway, like you did in laminas-session
1. Move `FormElementManagerV3Polyfill` to `FormElementManager` (already done)
1. Copy-paste `FormElementManagerTrait` content into `FormElementManager` with the purpose of simplify maintenance (already done)
1. Create a "new" `FormElementManagerV3Polyfill` as empty subclass of the new `FormElementManager`
1. Mark the "new" `FormElementManagerV3Polyfill` as `@deprecated`

This way we can:

1. Get rid of an old class
1. Get rid of `class_alias` burden
1. Refactor the code to be more easily maintainable
1. Keep BC

Right?

Reply:
@Slamdunk 
Very good solution! 👍 
_(Much better than my short list.)_

Reply:
Done